php/libraries/DefaultFilesUploadHandler.class.inc

```diff
@@ -0,0 +1,367 @@
+                                );
+                            } else {
+                                if ($this->_dirGroup != '') {
+                                    if (!chgrp($newDir, $this->_dirGroup)) {
+                                        $this->addError(
+                                            $field,
+                                            "Command 'chgrp {$this->_dirGroup} "
+                                                . "$newDir' failed"
+                                        );
+                                    }
+                                }
+                            }
+                            // File name exists. Make sure it refers to a directory.
+                        } else if (!is_dir($newDir)) {
+                            $this->addError(
+                                $field,
+                                "Cannot move {$file['tmp_name']} to $newDir: "
+                                    . "$newDir already exists and is not a directory"
+                            );
+                        }
+
+                        // If everything went OK so far
+                        if (!isset($this->_errors[$field])) {
+                            // Move the file to its final location
+                            $oldPath = $file['tmp_name'];
+                            $newPath = $newFile->getPathname();
+                            if (!move_uploaded_file($oldPath, $newPath)) {
+                                $this->addError(
+                                    $field,
+                                    "Could not move $oldPath to $newPath"
+                                );
+                            } else {
+                                $this->_movedFiles[$field] = $newFile;
+                            }
+                        }
+
+                        // Signal that an attempt to move a file from its temporary
+                        // location to its final location was made
+                        $this->onFileMoveEvent(
+                            $file['tmp_name'],
+                            isset($this->_errors[$field])
+                                ? null : $newFile->getRealPath(),
+                            $this->_errors[$field] ?? null
+                        );
+                    }
+                }
+            }
+        } finally {
+            $this->processingDone();
+        }
+    }
+
+    /**
+     * Finds the new (final, permanent) location for an uploaded file.
+     * This method only computes the new file path and does not actually
+     * move the file.
+     *
+     * @param string $tmpFilePath temporary path (on the server) where the
+     *               uploaded file is currently located.
+     *
+     * @return SplInfo the path information for the new file location.
+     */
+    public function getNewLocation(string $tmpFilePath) : SplFileInfo
```

Comment:
By passing the LorisFormFileElement instead of its tmp_name attribute, the foreach loop could be avoided. And this could instanciate the SPLFileInfo and check if it is writable before returning it. throw an exception maybe.